From be7e77080ff89a1de91953a47cff06ef3aa7c15c Mon Sep 17 00:00:00 2001
From: Lawrence Rust <lvr@softsystem.co.uk>
Date: Fri, 4 Mar 2016 11:21:37 +0000
Subject: [PATCH 306/348] MythMusic: Avoid a SEGV in the CD ripper ogg encoder

VorbisEncoder ctor calls ogg_packet_clear with an uninitialised
ogg_packet.  This can result in a call to free() with a random ptr.
It is preferable to simply zero the ogg_packet before use.

Also VorbisEncoder dtor calls vorbis_block_clear before
vorbis_dsp_clear which often results in a SEGV.  Reversing the
order of these calls prevents the SEGV.

This patch also removes some unused member variables.

Signed-off-by: Lawrence Rust <lvr@softsystem.co.uk>
---
 mythplugins/mythmusic/mythmusic/vorbisencoder.cpp |    8 +++++---
 mythplugins/mythmusic/mythmusic/vorbisencoder.h   |    2 --
 2 files changed, 5 insertions(+), 5 deletions(-)

diff --git a/mythplugins/mythmusic/mythmusic/vorbisencoder.cpp b/mythplugins/mythmusic/mythmusic/vorbisencoder.cpp
index bc8b7f4..3955e86 100644
--- a/mythplugins/mythmusic/mythmusic/vorbisencoder.cpp
+++ b/mythplugins/mythmusic/mythmusic/vorbisencoder.cpp
@@ -6,6 +6,7 @@
 
 // ANSI C
 #include <cstdlib>
+#include <cstring>
 
 // C++
 #include <iostream>
@@ -38,14 +39,13 @@ VorbisEncoder::VorbisEncoder(const QString &outfile, int qualitylevel,
                              MusicMetadata *metadata) :
     Encoder(outfile, qualitylevel, metadata),
     packetsdone(0),
-    eos(0),
     bytes_written(0L)
 {
     vorbis_comment_init(&vc);
 
     vorbis_info_init(&vi);
 
-    ogg_packet_clear(&op);
+    std::memset(&op, 0, sizeof op);
 
     float quality = 1.0;
     if (qualitylevel == 0)
@@ -80,6 +80,7 @@ VorbisEncoder::VorbisEncoder(const QString &outfile, int qualitylevel,
     ogg_stream_packetin(&os, &header_comments);
     ogg_stream_packetin(&os, &header_codebooks);
 
+    ogg_page og;
     int result;
     while ((result = ogg_stream_flush(&os, &og)))
     {
@@ -98,8 +99,8 @@ VorbisEncoder::~VorbisEncoder()
 {
     addSamples(0, 0); //flush
     ogg_stream_clear(&os);
-    vorbis_block_clear(&vb);
     vorbis_dsp_clear(&vd);
+    vorbis_block_clear(&vb);
     vorbis_comment_clear(&vc);
     vorbis_info_clear(&vi);
 
@@ -144,6 +145,7 @@ int VorbisEncoder::addSamples(int16_t * bytes, unsigned int length)
             int eos = 0;
             while (!eos)
             {
+                ogg_page og;
                 int result = ogg_stream_pageout(&os, &og);
                 if (!result)
                     break;
diff --git a/mythplugins/mythmusic/mythmusic/vorbisencoder.h b/mythplugins/mythmusic/mythmusic/vorbisencoder.h
index 724a95d..b8864d3 100644
--- a/mythplugins/mythmusic/mythmusic/vorbisencoder.h
+++ b/mythplugins/mythmusic/mythmusic/vorbisencoder.h
@@ -15,10 +15,8 @@ class VorbisEncoder : public Encoder
     int addSamples(int16_t *bytes, unsigned int len);
 
   private:
-    ogg_page og;
     ogg_packet op;
     long packetsdone;
-    int eos;
     long bytes_written;
     vorbis_comment vc;
     ogg_stream_state os;
-- 
1.7.9.5

